### PR TITLE
Add support for ephemeral document messages.

### PR DESCRIPTION
**Pull request by pvh**
Follows the progress API and the previous hypermerge version's design. Works in Pushpin. (See `upgrade` branch.)

There's a failing test which for some reason is receiving the message twice instead of once. No clue why yet but if you notice, let me know. There seem to be two copies of the peer to send the message to and I'm not sure why yet.

### CODE REVIEW

File: src/RepoBackend.ts [hunk 1]
```diff
@@ -345,6 +345,17 @@ export class RepoBackend {
         });
         break
       }
+      case "DocumentMessage": {
+        const { contents, id } = (msg as DocumentMsg);
+        this.meta.docsWith(id).forEach( (doc: string) => {
```
Comment (mjtognetti) [inline src/RepoBackend.ts]: Not sure if this is the right thing to do. Might need some further thought.

Comment (mjtognetti) [reply src/RepoBackend.ts]: This seems incorrect - supposedly we're trying to send a message to a particular document, not the set of documents this actor is participating in, each of which has its own document id. We should either only send messages to the document specified by `id` here, or change the message semantics to reflect the intent to send the message to all documents this actor is participating in. 

More of "actor broadcast"? Don't have a use case in mind for that atm.

File: src/Handle.ts [hunk 2]
```diff
@@ -82,8 +95,22 @@ export class Handle<T> {
     return this
   }
 
+  subscribeMessage = (
+    subscriber: (event: any) => void
+  ): this => {
+    if (this.messageSubscription) {
+      throw new Error("only one progress subscriber for a doc handle")
```
Comment (mjtognetti) [inline src/Handle.ts]: copy/paste. Should probably say "only one message subscriber for a doc handle"

File: src/Handle.ts [hunk 3]
```diff
@@ -45,6 +52,12 @@ export class Handle<T> {
     }
   }
 
+  pushMessage = (contents: any) => {
```
Comment (mjtognetti) [inline src/Handle.ts]: We really need a way to specify private/internal methods - or at least have a better naming scheme. `message()` vs `pushMessage`() is pretty ambiguous.

File: src/RepoBackend.ts [hunk 4]
```diff
@@ -555,6 +566,14 @@ export class RepoBackend {
           this.open(msg.id);
           break;
         }
+        case "DocumentMessage": {
+          const { id, contents } = msg
```
Comment (mjtognetti) [inline src/RepoBackend.ts]: This is a case where `id` confusingly ambiguous. If you think of it as a document id, the following code doesn't make sense. If you think of it as a document id, other parts of the code don't make sense!

Comment (jeffpeterson) [reply src/RepoBackend.ts]: I could be misreading, but it seems this is looking up the collaborating actors for a given doc id incorrectly. Maybe be something like:
```ts
const docBackend = this.docs.get(id)
if (docBackend) {
  const documentActor = this.actor(docBackend.actorId)
  // ...
}
```

I could imagine the `DocBackend` having a method that returns the collaborating actors to simplify this a bit.

File: src/RepoBackend.ts (hunk 1 above)
Comment (pvh) [reply src/RepoBackend.ts]: I have done this.

File: src/Handle.ts (hunk 3 above)
Comment (pvh) [reply src/Handle.ts]: I agree with this but I didn't do it. :D

File: src/RepoBackend.ts (hunk 4 above)
Comment (pvh) [reply src/RepoBackend.ts]: Fixed. Yeah, this stuff is kind of messy. I think using explicit actor & document ID types would help differentiate all this (and find places where we've conflated them.)

**Pull request merged**